include/lensfun/lensfun.h.in

```diff
@@ -1373,11 +1371,11 @@ struct LF_EXPORT lfLens
     private:
         void UpdateLegacyCalibPointers();
 
-        std::vector<lfLensCalibrationSet*> Calibrations;
+        std::vector<std::shared_ptr<lfLensCalibrationSet>> Calibrations;
```

Comment:
That make it incompatible with the following method.
https://github.com/lensfun/lensfun/blob/20cbe397dc8e11eea8ea0a7ac4ca42b2b2413f34/libs/lensfun/lens.cpp#L1315C44-L1315C62

And due to C-style reinterpret cast the problem got hidden.

Reply:
You are right, it's not a continous memory any longer. I'll improve the implementation.

Reply:
I've added a test case for this method, the thing is that we should return data to the user of the API. We can't create a new object here because lensfun is the owner of the data, and we can't return the internals of the std::vector any more because they are smart pointers now. 

I have created a data member which holds the C data structure. When Calibrations is updated we also update the C API data structure. 

```
lfLensCalibrationSet** calibrationsCAPI;

void lfLens::UpdateCalibrationsCAPI() {
    if(calibrationsCAPI != nullptr) {
        free(calibrationsCAPI);
        calibrationsCAPI = nullptr;
    }

    if(Calibrations.empty()) return;

    int sz = sizeof(*calibrationsCAPI) * (Calibrations.size()+1);
    calibrationsCAPI = (lfLensCalibrationSet**)malloc(sz);
    calibrationsCAPI[Calibrations.size()] = nullptr;

    lfLensCalibrationSet** it = calibrationsCAPI;
    for(const auto& c : Calibrations)
        *(it++) = c.get();
}

const lfLensCalibrationSet* const* lfLens::GetCalibrationSets() const
{
    return calibrationsCAPI;
}

```
I don't like this solution because we now introduce more complexity. Do we really need a list of Calibrations, because as far as I can see we only add one object to the list as far as I can see.